src/dmd/sideeffect.d

```diff
@@ -211,14 +211,67 @@ private bool lambdaHasSideEffect(Expression e)
     return false;
 }
 
+private bool isAssignment(Expression e)
+{
+    extern (C++) final class IsAssignment : SemanticTimePermissiveVisitor
+    {
+        alias visit = typeof(super).visit;
+        bool result;
+
+        override void visit(AssignExp)
+        {
+            result = true;
+        }
+
+        override void visit(BinAssignExp)
+        {
+            result = true;
+        }
+    }
+
+    scope IsAssignment v = new IsAssignment();
+    e.accept(v);
+    return v.result;
+}
+
+/**
+ * Check whether the e is a call to a @nodiscard function, or is
+ * a non-assignment expression with @nodiscard type.
+ */
```

Comment:
The comment should specify that this function issues an error in the above mentioned cases.

Also, this function is called "isNodiscard", therefore the expectation would be that it simply verifies if an expression contains an @nodiscard expression. However, this function does more than that: it also errors if the expression is nodiscard. I suggest you rename the function to checkNodiscard. This implies that an error might be issued.